src/api.ts

```diff
@@ -390,37 +400,73 @@ function createApi({
 
       const { port1, port2 } = new MessageChannel()
 
-      port2.on('message', (/** message: PortMessage */) => {
-        // TODO: do something with progress messages
+      activeImports.set(importId, port2)
+
+      piscina
+        .run(
+          {
+            dbPath: db.name,
+            importId,
+            mbTilesDbPath: mbTilesDb.name,
+            // TODO: `style` is not guaranteed to exist since the tileset could be a vector tileset
+            // and we don't generate a style for those on tileset creation yet.
+            // Absence presents various complications when creating and updating offline area and imports in db
+            styleId,
+            tilesetId,
+            port: port1,
+          },
+          { transferList: [port1] }
+        )
+        .catch((err) => {
+          port2.emit('error', err)
+        })
+        .finally(() => {
+          port2.close()
+          activeImports.delete(importId)
+        })
+
+      return new Promise((res, rej) => {
+        port2.on('message', handleFirstProgressMessage)
+        port2.on('error', (err) => rej(err))
+
```

Comment:
Maybe put the piscina.run() inside this promise, after the port2 handlers have been added. I don't _think_ the piscina promise should every catch on the same tick, but if it does, and we do `port2.emit('error')`, node considers an eventHandler emitting an event called `error` that does not have a listener attached as an uncaught error, and will terminate the process, so we should make sure that the error handler is attached before. Actually, if the piscina run is inside this promise we don't need the emit and listener, can just reject from the piscina catch.

Reply:
> Actually, if the piscina run is inside this promise we don't need the emit and listener,

just to clarify, don't we still need the `message` listener so that we can resolve on the first message that gets received? The point about the error listener makes sense though

Reply:
Yes, I meant just the error emit and listener. Look out for the 'error' emit issue with the Node `EventEmitter` - it's specific and unusual behaviour that is the message is called `error` and it is not handled then it throws an uncaught error. It has really frustrated me before, and it doesn't really make sense. Just one to keep an eye on with any code that uses an eventEmitter!